### ISSUE TITLE
Microphone sounds real bad when running on steamdeck

### ISSUE THREAD

**Issue by Quatsz** [Vencord/Vesktop#520]
The frequency or something is real off with just vesktop.. regular discord / chrome picks up the mic just fine. It's not the internal mic im using but the issue is there with that one too. When I test it, its like the bitrate was just really low. sounded like a mcdonalds happy meal toy. everything else works though! Real happy with being able to stream audio... just cant seem to make my microphone sound like it should on the steamdeck. tried everything i could from my end using settings, not very linux savvy so any help or suggestions i appreciated. the issue persists with both internal mic and using a headphone mic. thanks for reading!

**Comment by Curve** [comment]
I don't see how this would happen, microphone audio is not tinkered with (afaik) and other clients also just use electron

**Comment by gwstorm** [comment]
Agreed. Mine was more of a coincidental observation. Although the version I started with was using Electron29. I just swapped to 30 and it's sorted out... Not sure why. Maybe it was just the reinstall. I'll stop hijacking this post now.

**Comment by GTcreyon** [comment]
I have this problem too, running on my Linux laptop. People complain about a loud popping or crackling sound.

**Comment by VelamyGH** [comment]
having this issue too, was able to most likely confirm that theres a lot of compression going on using an audio in source to see what vesktop picks up.

this video shows what vesktop is (most likely) picking up for audio https://velamy.s-ul.eu/gPjFVgfi
this is the direct audio https://velamy.s-ul.eu/V8yav6KU

i have none of the noise suppression options turned on in discord settings, the quality is just that bad overall and i havent found a way to fix it. this is the most relevant issue i could find relating to this

**Comment by Vendicated** [comment]
have had multiple users say that they can't repro, so this seems to be a system issue. this issue will likely go stale since I have no steamdeck, help by someone with one would be appreciated

**Comment by VelamyGH** [comment]
should mention i came into this meaning to say this seems to be an issue on linux, or at the very least arch

**Comment by Curve** [comment]
> issue on linux in general, or at the very least arch and not just steamOS

I'm also running arch, as are multiple people I know and I know of no one that's experiencing a similar issue.

Maybe this issue is related to the specific microphone / microphone drivers used? It may be helpful to provide output of `pw-dump`, that could give us a lot of insights into the device used, as well as the audio settings used for it.

**Comment by VelamyGH** [comment]
>It may be helpful to provide output of `pw-dump`

getting the entire output of this gives way too many lines for my terminal, cant seem to locate or find a way to see the entire output. is there any specific info i should gather? or any way to get the entire JSON it outputs

**Comment by VelamyGH** [comment]
> @VelamyGH run `pw-dump > log.txt` then provide the log.txt file

thank you :) [log.txt](https://github.com/user-attachments/files/15919007/log.txt)

**Comment by Patpatolino** [comment]
Same problem for me, "official" discord application microphone input is totally fine.
Vesktop or other forks is the audio input quality kinda.. "muffled".
[
[log.txt](https://github.com/user-attachments/files/16036838/log.txt)
](url)

**Comment by VelamyGH** [comment]
small update to this, but i did a fresh install of endeavourOS on my laptop (separate hardware from my original post), and i still have the same mic issues persist. here is the log from a fresh install with no tweaking done to vesktop or pipewire 
[log.txt](https://github.com/user-attachments/files/16136742/log.txt)

**Comment by DerCommander323** [comment]
I daily drive the Steam Deck Oled docked, and I believe I also had this issue when electron29 was introduced to vesktop. I edited the launch script to use electron28, which seemed to resolve it. Now on electron31 I havent heard any of my friends complain yet either.

**Comment by SOLDATO2** [comment]
> Same problem for me, "official" discord application microphone input is totally fine. Vesktop or other forks is the audio input quality kinda.. "muffled".
> 
> [log.txt](https://github.com/user-attachments/files/16036838/log.txt)

I can also confirm I have this exact issue, I'm on manjaro. My voice sounds fine on Discord's official client but overall I've noticed a considerable loss in quality in Vencord.

My voice sounds muffled whenever I have krisp/standard noise suppression enabled. 


Also, even with nois sup disabled, I've noticed there's still some sort of loss in quality, it just sounds less "clean" as if there's still some "trash"  present (echoes, static, etc...).

**Comment by VelamyGH** [comment]
> I daily drive the Steam Deck Oled docked, and I believe I also had this issue when electron29 was introduced to vesktop. I edited the launch script to use electron28, which seemed to resolve it. Now on electron31 I havent heard any of my friends complain yet either.

would like to test this too, how exactly did you edit the launch script to do this?

**Comment by JoshuaCF** [comment]
I'd also like to say I have this issue, and I'm running an effectively unmodified Linux Mint Xfce installation. The only thing I could have messed up should be my graphics drivers, but the issue existed before that muck-up.

Using Vesktop without noise suppression, same as Discord, makes my microphone sound terrible to all the people I talk to. Turning on noise suppression brings it close to the original non-suppressed quality of the regular Discord client. Let me know what information I can provide to help with the issue.

**Comment by djsigmann** [comment]
I have this issue as well, I'm on arch, running pipewire and x11.
The same problem occurs in all other 3rd-party discord clients I've tried and the discord web-app running in both firefox and brave.

The official discord client works fine.

**Comment by TekkertheChaot** [comment]
I just switched from EndeavorOS (arch) to Bazzite (atomic fedora) and came upon this issue myself. 
The native Discord flatpak has no problem picking up my mic without cutoffs. Any other client (goofcord as well) has this problem. 
Someone recommended putting `--disable-features=WebRtcAllowInputVolumeAdjustment` into the launch options of the flatpak but that doesn't seem to work for me.

**Comment by Damglador** [comment]
My issue is a bit different. For me, the audio is crackling and cutting in and out a lot. Happens on Equibop (Vesktop fork) AUR and flatpak and Vesktop flatpak. In Zen Browser and the official client audio is fine.

**Comment by Damglador** [comment]
Probably related https://github.com/electron/electron/issues/36757

**Comment by Damglador** [comment]
After checking qpwgraph, Equibop uses "Chromium Input", while the official client uses "WEBRTC VoiceEngine". That's probably why the official client doesn't have the issue, while Vesktop, Equibop and WebCord (all I tested) do.

**Comment by djsigmann** [comment]
> After checking qpwgraph, Equibop uses "Chromium Input", while the official client uses "WEBRTC VoiceEngine". That's probably why the official client doesn't have the issue, while Vesktop, Equibop and WebCord (all I tested) do.
> 
> I've also tested Chromium flatpak and it has the same issue. The issue also happens only with Discord, other Chromium-based audio capture works fine

Out of curiosity, does this problem still occur for you if you use discord site on firefox?

**Comment by Vendicated** [comment]
Comparing to Firefox is meaningless. Compare it to chromium (not Chrome or any other fork) instead

**Comment by djsigmann** [comment]
> Comparing to Firefox is meaningless. Compare it to chromium (not Chrome or any other fork) instead

Well, I experience the same issue on Firefox, surely there's something to that?

**Comment by gator-git** [comment]
> No, it's exclusive to Discord on Chromium. At least on Zen I don't have the issue

Same here, I will check on normal chromium if that does make a difference.

**Comment by Damglador** [comment]
The issue magically went away for me

**Comment by djsigmann** [comment]
This issue seems to have been fixed for me as well.

**Comment by Kawaiidoscope** [comment]
I tested this yesterday, I could stream on Linux with Wayland just fine, but microphone quality is very bad. I switched to officiall discord app, and everyone told me sound quality is way better. I use the flatpak for both apps

**Comment by ReflectJudo** [comment]
I've suffered from this on various Linux installs over the last year or two including an install on Windows. It makes the microphone quality unbearably bad. The same thing happens just using web Discord period.

**Comment by SOLDATO2** [comment]
Yes, this seems specific to the vesktop client. This issue is happening on both windows & linux.

This issue isn't present on the official discord client + vencord patch.

**Comment by pinkyzet366** [comment]
Do you guys have any updates on this? 

I have the same problem here on Equibop and Vesktop, using fedora 42, the mic quality is really bad.
Ive noticed that the switchers in noise-cancelling thingies doesn't work, and in official discord client everything works

**Comment by Theoldnation** [comment]
Same problem here

**Comment by Megara18** [comment]
Been testing on NobaraOS (Fedora) more Modded clients and all of them haves this issue... Maybe is related to Electron itself?

(Not using SteamDeck, using a Desktop PC with Linux NobaraOS)

**Comment by Damglador** [comment]
@Megara18 in my testing this was an issue even in Chromium

**Comment by Vendicated** [comment]
Yes all these issues are chromium issues and there's not much we can do about them

**Comment by OrdinaryCloverOC** [comment]
it only happens when i plug in headphones for me. (bazzite) it seemingly happens randomly cause i open discord with no issues, then people complain about my mic as if it's awful an hour into conversation. i test my microphone in discord and it's fine. so many people have complained, but it appears to only be in call and NOT in the mic test.

**Comment by OrdinaryCloverOC** [comment]
i should mention the lcd model has issues with the headphone jack, so may be a hardware issue.

**Comment by troit5ky** [comment]
Please remove the “system issue” tag. The microphone cuts out in the desktop version (Flatpak) on Cachyos, but everything works fine in the original Discord app.

**Comment by ABrony** [comment]
Although i'm not on steamdeck, my friends too noticed change in sound of mic when i switched to vesktop. It's more muffled, as if it has lower bitrate and aggressive cutoff on higher frequencies. Also, it sometimes click and generally sounds worse. Native discord client is fine, both stable and canary. So it's certainly not mic problem.

Someone noted it's only in call, but for me it's also same in mic test

Would be cool if fix is possible. Overall vesktop works a bit better, especially when it comes to screen sharing

**Comment by VeryEvilHumna** [comment]
I think the issue is because Discord's echo cancellation sucks and can't be turned off: Discord web is probably completely ignoring the options in Voice & Video menu regarding voice processing (at least I couldn't find any differences with all the combinations). I was able to reproduce that in Discord web on Zen browser

I was able to fix my issue of a muffled sound + no high frequencies by forcing `echoCancellation`/`noiseSuppression`/`autoGainControl` `=false` and stripping `advanced/goog*` constraints in [src/renderer/patches/fixStreamConstraints.ts:12](https://github.com/Vencord/Vesktop/blob/cb9c55fe1aa4cd1d3fbe23bddfdca02627c8c372/src/renderer/patches/fixStreamConstraints.ts#L12) (getUserMedia + applyConstraints)

Patch: 
```patch
diff --git a/src/renderer/patches/fixStreamConstraints.ts b/src/renderer/patches/fixStreamConstraints.ts
index 04820a4..c15b59b 100644
--- a/src/renderer/patches/fixStreamConstraints.ts
+++ b/src/renderer/patches/fixStreamConstraints.ts
@@ -2,17 +2,42 @@
  * Vesktop, a desktop app aiming to give you a snappier Discord Experience
  * Copyright (c) 2025 Vendicated and Vesktop contributors
  * SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ * VoiceFix: Discord web ignores Voice & Video toggles and always sends
+ * echoCancellation/noiseSuppression/autoGainControl + advanced:[{autoGainControl}]
+ * Force-off regardless of MediaEngineStore - web equivalent of WEBRTC VoiceEngine.
  */
 
 import { Logger } from "@vencord/types/utils";
-import { MediaEngineStore } from "@vencord/types/webpack/common";
 
-const logger = new Logger("VesktopStreamFixes");
+const logger = new Logger("VesktopVoiceFix");
 
 function fixAudioTrackConstraints(constraint: MediaTrackConstraints) {
-    const target = constraint.advanced?.find(opt => Object.hasOwn(opt, "autoGainControl")) ?? constraint;
-
-    target.autoGainControl = MediaEngineStore.getAutomaticGainControl();
+    (constraint as any).echoCancellation = false;
+    (constraint as any).noiseSuppression = false;
+    (constraint as any).autoGainControl = false;
+    delete (constraint as any).googEchoCancellation;
+    delete (constraint as any).googEchoCancellation2;
+    delete (constraint as any).googNoiseSuppression;
+    delete (constraint as any).googNoiseSuppression2;
+    delete (constraint as any).googAutoGainControl;
+    delete (constraint as any).googAutoGainControl2;
+    delete (constraint as any).googHighpassFilter;
+    if ((constraint as any).advanced) {
+        const adv = (constraint as any).advanced as any[];
+        const filtered = adv.filter(
+            (o: any) =>
+                !("autoGainControl" in o) &&
+                !("echoCancellation" in o) &&
+                !("noiseSuppression" in o) &&
+                !("googAutoGainControl" in o) &&
+                !("googEchoCancellation" in o) &&
+                !("googNoiseSuppression" in o)
+        );
+        if (filtered.length === 0) delete (constraint as any).advanced;
+        else (constraint as any).advanced = filtered;
+    }
 }
 
 function fixVideoTrackConstraints(constraint: MediaTrackConstraints) {
@@ -23,44 +48,36 @@ function fixVideoTrackConstraints(constraint: MediaTrackConstraints) {
 
 function fixStreamConstraints(constraints: MediaStreamConstraints | undefined) {
     if (!constraints) return;
-
     if (constraints.audio) {
-        if (typeof constraints.audio !== "object") {
-            constraints.audio = {};
-        }
-        fixAudioTrackConstraints(constraints.audio);
+        if (typeof constraints.audio !== "object") constraints.audio = {};
+        if ((constraints as any).audio === true) (constraints as any).audio = {};
+        fixAudioTrackConstraints(constraints.audio as MediaTrackConstraints);
     }
-
     if (constraints.video) {
-        if (typeof constraints.video !== "object") {
-            constraints.video = {};
-        }
-        fixVideoTrackConstraints(constraints.video);
+        if (typeof constraints.video !== "object") constraints.video = {};
+        fixVideoTrackConstraints(constraints.video as MediaTrackConstraints);
     }
 }
 
 const originalGetUserMedia = navigator.mediaDevices.getUserMedia;
 navigator.mediaDevices.getUserMedia = function (constraints) {
     try {
-        fixStreamConstraints(constraints);
+        fixStreamConstraints(constraints as any);
     } catch (e) {
         logger.error("Failed to fix getUserMedia constraints", e);
     }
-    return originalGetUserMedia.call(this, constraints);
+    return originalGetUserMedia.call(this, constraints as any);
 };
 
 const originalApplyConstraints = MediaStreamTrack.prototype.applyConstraints;
 MediaStreamTrack.prototype.applyConstraints = function (constraints) {
     if (constraints) {
         try {
-            if (this.kind === "audio") {
-                fixAudioTrackConstraints(constraints);
-            } else if (this.kind === "video") {
-                fixVideoTrackConstraints(constraints);
-            }
+            if (this.kind === "audio") fixAudioTrackConstraints(constraints as MediaTrackConstraints);
+            else if (this.kind === "video") fixVideoTrackConstraints(constraints as MediaTrackConstraints);
         } catch (e) {
             logger.error("Failed to fix constraints", e);
         }
     }
-    return originalApplyConstraints.call(this, constraints);
+    return originalApplyConstraints.call(this, constraints as any);
 };

```

**Comment by troit5ky** [comment]
@VeryEvilHumna it works, thx!

**Comment by srn2202** [comment]
Can confirm, the fix by @VeryEvilHumna works wonders on Windows 10! Thank you so much, it has been bothering me for a long while now.

**Comment by VeryEvilHumna** [comment]
I guess it's time to wrap the fix into a plugin? Or make a PR...
